Add Entry General: Middleware (#599)
Codecademy/docs: content/general/concepts/middleware/middleware.md

---
Title: 'Middleware'

> Description: 'Middleware is software that connects applications together. Middleware connects data, APIs, software tools and other applications in other to bring a complete application to the end user.'

Reviewer: ```suggestion
Description: 'Middleware is software that connects data, APIs, software tools, and other applications in order to bring a complete application to the end user.'
```

Subjects:
  - 'Computer Science'
  - 'Web Development'
Tags:
  - 'APIs'
  - 'Database'
  - 'Middleware'

> CatalogContent: # Please use course/path landing page slugs, rather than linking to individual content items. If listing multiple items, please put the most relevant one first

Reviewer: ```suggestion
CatalogContent:
```

Those comments are just meant for templating purposes.

>   - 'learn-html'

Reviewer: Is there another `learn-` slug that relates to middleware more closely?

  - 'paths/back-end-engineer-career-path'
---

> Middleware is software that connects applications together. The term _middleware_ comes from the fact that this software sits between the front-end or client side of an application and the back-end where the resources are being requested. Middleware connects or "glues" together data, APIs, software tools and other applications in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.

Reviewer: ```suggestion
Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end or client side of an application and the back-end where the resources are being requested. Middleware connects or "glues" together data, APIs, software tools and other applications in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.
```

Most forms of emphasis are generally advised against. ☹️

Reviewer 2: Oh, yeah ☹️ no italics for future reference

## Types of Middleware

> ### Message-oriented middleware

Reviewer: ```suggestion
### Message Oriented Middleware
```
For headers, let's use strict "Title Case" 😄

Reviewer 2: AP Style* Title Cased

> This type of middleware allows applications to exchange messages and communicate between eachother.

Reviewer: ```suggestion
A message oriented middleware, or MOM, allows applications to exchange messages and communicate between one another.
```

Could include a short list of prominent examples for each of these middleware types?

### Database middleware

This type of middleware gives access to databases and provides points of interaction between a variety of databases.

> ### Application Programming Interface (API) Middleware

Reviewer: ```suggestion
### Application Programming Interface
```

---
Title: 'Middleware'
Description: 'Middleware is software that connects data, APIs, software tools, and other applications in order to bring a complete application to the end user.'
Subjects:
  - 'Computer Science'
  - 'Web Development'
Tags:
  - 'APIs'
  - 'Database'
  - 'Middleware'
CatalogContent:
  - 'learn-sql'
  - 'paths/back-end-engineer-career-path'
---

Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end, or client-side, of an application and the back-end, or server-side, where the resources are being requested.

> Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.

Reviewer 2: Thinking the last two sentences here can be shifted to after the "Types of Middleware" heading and before "Message Oriented Middleware"

## Types of Middleware

### Message Oriented Middleware

A message oriented middleware, or MOM, allows applications to exchange messages and communicate between one another. This middleware offers ease of use but can be costly at larger scales.

#### Examples

- RabbitMQ

RabbitMQ is a messaging broker platform used by applications to send and recieve messages.

- IBM MQ

IBM MQ is another example of message oriented middleware. Some of the core components include messages, message queues and queue manager.

- Amazon SNS

Amazon simple notification service is a way to manage messaging between applications and users.

### Database middleware

This type of middleware gives access to databases and provides points of interaction between a variety of databases. This middleware improves efficiency however it can be language specific which can create limitations.

#### Examples

- ODBC

> Open Database Connectivity is an API used for accessing database management systems.

Reviewer 2: Can we try placing these in a toggle. When we click on the toggle, it should expand to show the additional info on the middleware.

```
<details>
  <summary>ODBC</summary>
  Open Database Connectivity is an API used for accessing database management systems.
</details>
```

Author: > Can we try placing these in a toggle. When we click on the toggle, it should expand to show the additional info on the middleware.
> 
> ```
> <details>
>   <summary>ODBC</summary>
>   Open Database Connectivity is an API used for accessing database management systems.
> </details>
> ```

Should I do it for all the examples ? or just this one ?

---
Title: 'Middleware'
Description: 'Middleware is software that connects data, APIs, software tools, and other applications in order to bring a complete application to the end user.'
Subjects:
  - 'Computer Science'
  - 'Web Development'
Tags:
  - 'APIs'
  - 'Database'
  - 'Middleware'
CatalogContent:
  - 'learn-sql'
  - 'paths/back-end-engineer-career-path'
---

Middleware is software that connects applications together. The term "middleware" comes from the fact that this software sits between the front-end, or client-side, of an application and the back-end, or server-side, where the resources are being requested.

## Types of Middleware

> Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware is listed below.

Reviewer 3: ```suggestion
Data, APIs, software tools, and other applications are connected by middleware in order to bring a complete or bundled application to the end user. There are many types of middleware software depending on the use cases. Some of the most commonly used middleware are listed below.
```

### Message Oriented Middleware

> A message oriented middleware, or MOM, allows applications to exchange messages and communicate between one another. This middleware offers ease of use but can be costly at larger scales.

Reviewer 3: ```suggestion
A message oriented middleware, or MOM, allows applications to exchange messages and communicate with one another. This middleware offers ease of use but can be costly at larger scales.
```

#### Examples

<details>
  <summary>RabbitMQ</summary>
RabbitMQ is a messaging broker platform used by applications to send and recieve messages.
</details>

<details>

>   <summary>IBM MQ</summary>
> IBM MQ is another example of message oriented middleware. Some of the core components include messages, message queues and queue manager.

Reviewer: ```suggestion
  <summary>IBM MQ</summary>
  IBM MQ is another example of message oriented middleware. Some of the core components include messages, message queues and queue manager.
```

</details>

<details>
  <summary>Amazon SNS</summary>
Amazon simple notification service is a way to manage messaging between applications and users.
</details>

### Database Middleware

This type of middleware gives access to databases and provides points of interaction between a variety of databases. This middleware improves efficiency however it can be language specific which can create limitations.

#### Examples

<details>
  <summary>ODBC</summary>
  Open Database Connectivity is an API used for accessing database management systems.
</details>

<details>
  <summary>JDBC</summary>
Java Database Connectivity is an API that utilizes the Java programming language to access databases.
</details>

### API Middleware

> This type of middleware provides developers with the tools to create and manage APIs used in applications. This middleware offers scalability but connections can be slow when dealing with different servers.

Reviewer: ```suggestion
This type of middleware provides developers with the tools to create and manage APIs used in applications. While it offers scalability, connections can be slow when dealing with different servers.
```

#### Examples

<details>
  <summary>API management platforms</summary>

> API management platforms allow developers to access, control, and analyze APIs.

Reviewer 2: ```suggestion
  API management platforms allow developers to access, control, and analyze APIs.
```

</details>

<details>
  <summary>API gateways</summary>

>   API gateways provide services related to user authentication and rate litmiting.

Reviewer 3: ```suggestion
  API gateways provide services related to user authentication and rate limiting.
```

> API gateways provide services related to user authentication and rate litmiting.

Reviewer 2: ```suggestion
  API gateways provide services related to user authentication and rate litmiting.
```
